---
bibliography: manuscript.bib
csl: systematic-biology.csl
site: "bookdown::bookdown_site"
output:
  bookdown::pdf_document2:
    toc: FALSE
indent: TRUE
monofont: "Times New Roman"
fontsize: 12pt
language:
  label:
    fig: "**Figure S**"
    tab: "**Table S**"
header-includes:
  - \usepackage[left]{lineno}
  - \linenumbers
  - \usepackage{setspace}
  - \doublespacing
---

```{r setup, include=FALSE}
	# Load packages
	## General
library(knitr)
library(tidyverse)
library(stringr)
library(reshape2)
	## Biological
library(Rphylip)
library(arbutus)
library(vegan)
library(ape)
library(phangorn)
library(phytools)
library(OUwie)
library(picante)
library(geiger)
library(phylobase)
library(fields)
library(phylosignal)
library(geomorph)
library(phylopath)
library(phylolm)
	## Graphics
library(FactoMineR)
library(factoextra)
library(corrplot)
library(BAMMtools)
library(gridExtra)
library(xtable)
library(colorRamps)

	# Configure knitr, see http://yihui.name/knitr/options
	opts_knit$set(
	  progress=TRUE,
	  verbose=TRUE)
	opts_chunk$set(
	#  include=FALSE,
	  cache=TRUE,
	  echo=FALSE,
	  message=FALSE
	 )
knitr::opts_chunk$set(echo = TRUE)

```

# Supplementary Materials File {-}

Alejandro Damian-Serrano, Steven H.D. Haddock, Casey W. Dunn

## Supplementary Methods {-}

DAPC optimization: Some taxa have inapplicable states for certain absent characters (such as the length of a nematocyst subtype that is not present in a species), which are problematic for DAPC analyses. We tackled this by transforming the absent states to zeroes. This approach allows us to incorporate all the data, but creates an attraction bias between small character states (*e.g.* small tentilla) and absent states (*e.g.* no tentilla). Absent characters are likely to be very biologically relevant to prey capture and we believe they should be accounted for in a predictive approach. We limited the number of linear discriminant functions retained to the number of groupings in each case. We selected the number of principal components retained using the a-score optimization function (R adegenet::optim.a.score) [@jombart2010discriminant] with 100 iterations, which yielded more stable results than the cross validation function (R adegenet::xval). This optimization aims to find the compromise value with highest discrimination power with the least overfitting.

In order to explore the correlational structure among continuous characters and among their evolutionary histories, we used principal component analysis (PCA) and phylogenetic PCA [@revell2012phytools]. Since the character dataset contains many gaps due to missing characters and inapplicable states, we carried out these analyses on a subset of species and characters that allowed for the most complete dataset. This was done by removing the terminal filament characters (which are only shared by a small subset of species), and then removing species which had inapplicable states for the remaining characters. In addition, we obtained the correlations between the phylogenetic independent contrasts [@felsenstein1985phylogenies] using the package rphylip [@revell2014rphylip]. 

## Supplementary Results {-}

![\label{figure3} Heatmap summarizing the morphological diversity measured for 96 species of siphonophores clustered by similarity (raw data in Dryad repository). Missing values from absent characters presented as dark grey cells, missing values produced from technical difficulties presented as white cells. Values scaled by character.](Figures_final/figure3.jpg)

*Generating dietary hypotheses using tentillum morphology* – When predicting soft- and hard-bodied prey specialization, the DAPC achieved 90.9% discrimination success, only marginally confounding hard-bodied specialists with generalists (SM\@ref(DAPCsofthard)). The main characters driving this discrimination are involucrum length, heteroneme number, heteroneme volume, tentacle width, total nematocyst volume, total haploneme volume, elastic strand width, and heteroneme length. Discriminant analyses and GLM logistic regressions were also applied to specific prey type presence and selectivity (Table 2 int he main text), revealing the sign of their predictive relationship to each prey type. We only selected prey types with sufficient variation in the data to carry out these analyses (copepods, fish, and large crustaceans). While the presence of fish or large crustaceans in the diet cannot be unambiguously discriminated using tentillum morphology (SM\@ref(DAPCfish),SM\@ref(DAPCcrustacean)), specialization on fish or large crustacean prey can be fully disentangled (SM\@ref(DPACguilds)). For each prey type studied, tentillum morphology is a much better predictor of prey selectivity than of prey presence in the diet, despite prey selectivity data being available for a smaller subset of species. Interestingly, many of the morphological predictors had opposite slope signs when predicting prey selectivity *versus* predicting prey presence in the diet (Table 2 in the main text).
  
  ![\label{figure6} Hypothetical feeding guilds for siphonophore species predicted by a 6 PCA DAPC. Cell darkness indicates posterior probability of belonging to each guild. Training data set transformed so inapplicable states are computed as zeroes. Species ordered and colored according to their predicted feeding guild.](Figures_final/figure6.jpg)
  
*Evolution of tentillum and nematocyst characters* – One third of the characters measured support a non-phylogenetic generative model, indicating they are not phylogenetically conserved (SM\@ref(ModelSupport)). Total nematocyst volume and cnidoband-to-heteroneme length ratio showed strongly conserved phylogenetic signals. 74% of characters present a significant phylogenetic signal, yet only total nematocyst volume, haploneme length, and heteroneme-to-cnidoband length ratio had a phylogenetic signal with K > 1. 67% of characters support BM models, indicating a history of neutral constant divergence. No relationship between phylogenetic signal and BM model support was found. Haploneme nematocyst length is the only character with support for an EB model of decreasing rate of evolution with time. No character had support for a single-optimum OU model (when uninformed by feeding guild regime priors).

![\label{figure8} A. Correlogram showing strength of ordinary (upper triangle) and phylogenetic (lower triangle) correlations between characters. Both size and color of the circles indicate the strength of the correlation (R^2^). B. Scatterplot of phylogenetic correlation against ordinary correlation showing a strong linear relationship (R^2^ = 0.92, 95% confidence between 0.90 and 0.93). Light red and blue boxes indicate congruent negative and positive correlations respectively. Darker red and blue boxes indicate strong (<-0.5 or >0.5) negative and positive correlation coefficients respectively.](Figures_final/figure8.jpg){height=600px}

*Phenotypic integration of the tentillum* – Of the phylogenetic correlations (SM\@ref(figure8)a, lower triangle), 81.3% were positive and 18.7% were negative, while of the ordinary correlations (SM\@ref(figure8)a, upper triangle) 74.6% were positive and 25.4% were negative. Half (49.9%) of phylogenetic correlations were >0.5, while only 3.6% are < -0.5. Similarly, of the across-species correlations, 49.1% were >0.5 and only 1.5% were < -0.5. We found that 13.9% of character pairs had opposing phylogenetic and ordinary correlation coefficients. Just 4% have negative phylogenetic and positive ordinary correlations (such as rhopaloneme elongation ~ heteroneme-to-cnidoband length ratio and haploneme elongation, or haploneme elongation ~ heteroneme number), and only 9.9% of character pairs had positive phylogenetic correlation yet negative ordinary correlation (such as heteroneme elongation ~ cnidoband convolution and involucrum length, or rhopaloneme elongation with cnidoband length). These disparities can be caused by Simpson’s paradox [@blyth1972simpson]: the reversal of the sign of a relationship when a third variable (or a phylogenetic topology [@uyeda2018rethinking]) is considered. However, no character pair had correlation coefficient differences larger than 0.64 between ordinary and phylogenetic correlations (heteroneme shaft extension ~ rhopaloneme elongation has a Pearson’s correlation of 0.10 and a phylogenetic correlation of  -0.54). Rhopaloneme elongation shows the most incongruencies between phylogenetic and ordinary correlations with other characters.


![\label{figure9} Phylomorphospace of the simple continuous characters principal components, excluding ratios and composite characters. A. Variance explained by each variable in the PC1-PC2 plane. Axis labels include the phylogenetic signal (K) for each component and p-value. B. Phylogenetic relationships between the species points distributed in that same space.](Figures_final/figure9.jpg){height=600px}

  In the non-phylogenetic PCA morphospace using only simple characters (SM\@ref(figure9)), PC1 (aligned with tentillum and tentacle size) explained 69.3% of the variation in the tentillum morphospace, whereas PC2 (aligned with heteroneme length, heteroneme number, and haploneme arrangement) explained 13.5%. In a phylogenetic PCA, 63% of the evolutionary variation in the morphospace is explained by PC1 (aligned with shifts in tentillum size), while 18% is explained by PC2 (aligned with shifts in heteroneme number and involucrum length). 
  
*Functional morphology of tentillum and nematocyst discharge* –  Tentillum and nematocyst discharge high speed measurements are available in the Dryad repository. While the sample sizes of these measurements were insufficient to draw reliable statistical results at a phylogenetic level, we did observe patterns that may be relevant to their functional morphology. For example, cnidoband length is strongly correlated with discharge speed (p value = 0.0002). This is probably the sole driver of the considerable difference between euphysonect and calycophoran tentilla discharge speeds (average discharge speeds: 225.0mm/s and 41.8mm/s respectively; t-test p value = 0.011), since the euphysonects have larger tentilla than the calycophorans among the species recorded.

  We also observed that calycophoran haploneme tubules fire faster than those of euphysonects (T-test p value = 0.001). Haploneme nematocysts discharge 2.8x faster than heteroneme nematocysts (T-test p value = 0.0012). Finally, we observed that the stenoteles of the Euphysonectae discharge a helical filament that “drills” itself through the medium it penetrates as it everts.
  
## Supplementary Discussion {-}

*Generating hypotheses on siphonophore feeding ecology* – One motivation for our research was to understand the links between predator capture tools and their diets so we can generate hypotheses about the diets of siphonophores based on morphological characteristics. Indeed, our discriminant analyses were able to distinguish between different siphonophore diets based on morphological characters alone. The models produced by these analyses generated testable predictions about the diets of many species for which we only have morphological data of their tentacles. While the limited dataset used here is informative for generating tentative hypotheses, the empirical dietary data are still scarce and insufficient to cast robust predictions. This reveals the need to extensively characterize siphonophore diets and feeding habits. In future work, we can test these ecological hypotheses and validate these models by directly characterizing the diets of some of those siphonophore species. Predicting diet using morphology is a powerful tool to reconstruct food web topologies from community composition alone. In many of the ecological models found in the literature, interactions among the oceanic zooplankton have been treated as a black box [@mitra2009closure]. The ability to predict such interactions, including those of siphonophores and their prey, will enhance the taxonomic resolution of nutrient-flow models constructed from plankton community composition data.

*Evolutionary history of tentillum morphology* –This study produced the most speciose siphonophore molecular phylogeny to date, while incorporating the most recent findings in siphonophore deep node relationships. This phylogeny revealed for the first time that the genus *Erenna* is the sister to *Stephanomia amphytridis*. *Erenna* and *Stephanomia* bear the largest tentilla among all siphonophores, thus their monophyly indicates that there was a single evolutionary transition to giant tentilla. Siphonophore tentilla range in size from ~30 µm in some *Cordagalma* specimens to 2-4 cm in *Erenna* species, and up to 8 cm in *Stephanomia amphytridis* [@pugh2014review]. Most siphonophore tentilla measure between 175 and 1007 µm (1st and 3rd quartiles), with a median of 373 µm. The extreme gain of tentillum size in this newly found clade may have important implications for access to large prey size classes.

  Siphonophore tentilla are defined as lateral, monostichous evaginations of the tentacle gastrovascular lumen with epidermal nematocysts [@totton1965synopsis]. The buttons on *Physalia* tentacles were not traditionally regarded as tentilla, but [@bardi2007taxonomic] and our observations [@munro2018improved], confirm that the buttons contain evaginations of the gastrovascular lumen, thus satisfying all the criteria for the definition. In this light, and given that most Cystonectae bear conspicuous tentilla, we conclude (in agreement with [@munro2018improved]) that tentilla are likely ancestral to all siphonophores, and secondarily lost in *Apolemia* and *Bathyphysa conifera*.
  
  The clade Tendiculophora contains far more species than its relatives Cystonectae, Apolemiidae, and Pyrostephidae. An increase in clade richness and ecological diversification can be triggered by a ‘key innovation’ [@simpson1955major]. The evolutionary innovation of the Tendiculophora tentilla with shooting cnidobands and modular regions may have facilitated further dietary diversification.

\newpage

## Supplementary Materials {-}

![\label{appendix2} Character definitions. ](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/Appendix2.png){height=500px}


Other trees inferred in this work include:

![\label{MLIqtree} Maximum likelihood IQTree inference, unconstrained. Node labels are bootstrap support values.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/MLunconstrained.png){height=300px}

The constrained tree inferences utilized the following constrain topology:

![\label{Transcriptome_constraint} Topology used to constrain analyses (minimal topological statements based on the incongruences between the unconstrained tree and Munro et al. (2018). ](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/constrain.png){height=400px}

![\label{ML_constrained} Constrained IQTree ML inference. Node labels are bootstrap support values.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/MLconstrained.png){height=300px}

![\label{Bayes_unconstrained} Unconstrained Bayesian topology inference in RevBayes (node labels are Bayesian posteriors).](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/BAYESunconstrained.png){height=300px}

![\label{Bayes_constrained} Clade constrained Bayesian inference in RevBayes (node labels are Bayesian posteriors).](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/BayesConstrained.png){height=300px}

![\label{unconstrained_timetree} Unconstrained ultrametric Bayesian time tree branch length and topology inference in RevBayes (node labels are Bayesian posteriors). Arbitrary rooting.).](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/Unconstrained timetree.png){height=400px}

![\label{constrained_timetree} Ultrametric Bayesian time tree branch length inference in RevBayes (node labels are bayesian posteriors). Topology clamped to the Bayesian constrained topology inference in \@ref{Bayes_constrained}. Tree rooted using outgroup constraint.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/constrainedTimetree.png){height=500px}

\newpage

![\label{ModelSupport} Model support (delta AICc), phylogenetic signal (Blomberg’s K), and phylogenetic signal permutation test p-value for each continuous character. Ntaxa = Number of taxa used in the analyses after removing those where the character sate is inapplicable or the data is missing.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/modelSupport.png){height=600px}

![\label{ModelAdequacy} Model adequacy scores for the best model supported for each morphological character. Cvar = coefficient of variation of the absolute value of the contrasts. Svar = Slope of a linear model fitted to the absolute value of the contrasts against their expected variances. Sasr = slope of the contrasts against the ancestral state inferred at each corresponding node. Shgt = slope of the contrasts against node depth. Dcfd = Kolmolgorov-Smirnov D-statistic comparing contrasts to a normal distribution with SD equal to the root of the mean of squared contrasts. ](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/modelAdequacy.png)

![\label{ModelSupportOUwie} Model support (delta AICc) for each morphological character analyzed on the feeding guild reconstruction regime tree. OU1 = Single-optimum Ornstein-Uhlenbeck. OUm = Multi-optima Ornstein-Uhlenbeck. Model adequacy scores calculated for the best supported model only. Msig = mean of squared contrasts. Cvar = coefficient of variation of the absolute value of the contrasts. Svar = Slope of a linear model fitted to the absolute value of the contrasts against their expected variances. Sasr = slope of the contrasts against the ancestral state inferred at each corresponding node. Shgt = slope of the contrasts against node depth. Dcfd = Kolmolgorov-Smirnov D-statistic comparing contrasts to a normal distribution with SD equal to the root of the mean of squared contrasts.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/OUwie_modelsupport.png)

\newpage

DAPC discriminant analyses:

![\label{DAPCguilds} DAPC for Feeding guilds. Six PCs retained after a-score optimization (100 iterations). Four LDA functions used. Discriminant power on training set: 100%. Prediction posterior distribution heat map in main text Figure 6. Variable contribution (top quartile) calculated by the sum of the LDA variable loadings weighted by the eigenvalue of each LDA.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/DAPC_guilds.png){height=500px}

![\label{DAPCcopepods} DAPC for copepod presence in the diet. Eight PCs retained after a-score optimization (100 iterations). One LDA functions used. Discriminant power on training set: 95.4%. Grayscale heat map shows the posterior probability distribution of the predictions. Variable contribution (top quartile) calculated by the sum of the LDA variable loadings weighted by the eigenvalue of each LDA.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/DAPC_copepod.png)

![\label{DAPCfish} DAPC for fish presence in the diet. Three PCs retained after a-score optimization (100 iterations). One LDA function used. Discriminant power on training set: 68.1%. Grayscale heat map shows the posterior probability distribution of the predictions. Variable contribution (top quartile) calculated by the sum of the LDA variable loadings weighted by the eigenvalue of each LDA.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/DAPC_fish.png)

![\label{DAPCcrustacean} DAPC for large crustacean presence in the diet. Four PCs retained after a-score optimization (100 iterations). One LDA function used. Discriminant power on training set: 81.8%. Grayscale heat map shows the posterior probability distribution of the predictions. Variable contribution (top quartile) calculated by the sum of the LDA variable loadings weighted by the eigenvalue of each LDA.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/DAPC_crustacean.png)

![\label{DAPCsofthard} DAPC for soft-bodied vs. hard bodied prey specialization. Six PCs retained after a-score optimization (100 iterations). Two LDA functions used. Discriminant power on training set: 90.9%. Grayscale heat map shows the posterior probability distribution of the predictions. Variable contribution (top quartile) calculated by the sum of the LDA variable loadings weighted by the eigenvalue of each LDA.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/DAPC_softhard.png)

\newpage

Ordinary and phylogenetic logistinc regression of morphological characters and the presence/absence of prey types:

![\label{GLMs} Logistic regressions between continuous morphological characters and prey type presences. Ntaxa = number of taxa used in the analyses after removing taxa with missing diet data and inapplicable character states. phyloGLM = Phylogenetic generalized logistic regression model. GLM = Generalized logistic regression model. P = p-value. b = slope. Only cases with significant GLM fits were retained. Cells colored blue indicate phyloGLM p-value < 0.05. Cells colored green indicate GLM p-value < 0.05](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/PhyloGLM.png){height=300px}

\newpage

SIMMAP ancestral reconstructions of categorical characters:

![\label{SIMMAPtentilla} SIMMAP Tentilla presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/tentillaPresenceSIM.png)

![\label{SIMMAPproxhet} SIMMAP Cnidoband proximal heteroneme presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/proxhet_presence.png)

![\label{SIMMAPdesmorhopalo} SIMMAP Desmoneme+Rhopaloneme presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/desmorhopalo_SIM.png)

![\label{SIMMAPdischarge} SIMMAP Actively discharging cnidobands presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/discharge_SIM.png)

![\label{SIMMAP_ES} SIMMAP Elastic strands presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/Elasticstrands_SIM.png)

![\label{SIMMAP_distalDesmo} SIMMAP Cnidoband distal desmonemes presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/distalDesmo_SIM.png)

![\label{SIMMAP_Coiled} SIMMAP Coiled cnidoaband phenotype presence/absence.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/Coiled_Sim.png)

![\label{SIMMAP_HetType} SIMMAP Heteroneme type.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/hettype_SIM.png)

![\label{SIMMAP_HapType} SIMMAP Haploneme type.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/haptype_SIM.png)

![\label{SIMMAPguilds} SIMMAP Feeding guilds.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/guild_SIM.png)

\newpage

Bayesian Analysis of Macroevolutionary Mixtures (BAMM) results:

![\label{BAMM_HetElon} Siphonophore ultrametric tree with estimated trait diversification rates of heteroneme elongation, estimated by BAMM. Yellow = Rate acceleration. Purple = Rate slowdown.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/HetElon_BAMM.png)

![\label{BAMM_HetElon_cred} Bayesian 95% credible rate regime shift set for heteroneme elongation. Red = rate acceleration. Blue = rate slowdown. f values are the fraction of the probability of the data accounted for.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/HetElon_Cred.png)


![\label{BAMM_HapElon} Siphonophore ultrametric tree with estimated trait diversification rates of haploneme elongation, estimated by BAMM. Yellow = Rate acceleration. Purple = Rate slowdown.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/HapElon_BAMM.png)

![\label{BAMM_HapElon_cred} Bayesian 95% credible rate regime shift set for haploneme elongation. Red = rate acceleration. Blue = rate slowdown. f values are the fraction of the probability of the data accounted for.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/HapElon_Cred.png)

\newpage

Variance-Covariance Matrix analyses of phenotypic intergation and evolutionary modularity:

![\label{VCV_NAZero} Rate covariance matrix for the whole tree using all taxa (45 species), transforming inapplicable states to zeroes. Covariances scaled to correlations. All characters estimated simultaneously under Brownian Motion.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCV_Nazero.png)

![\label{VCV_NAremove} Rate covariance matrix for the whole tree using only taxa without inapplicable states (24 species). Covariances scaled to correlations. All characters estimated simultaneously under Brownian Motion.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCV_NAremove.png)

![\label{VCV_dietaxa} Rate covariance matrix for the whole tree using only taxa with diet data (22 species), transforming inapplicable states to zeroes. Covariances scaled to correlations. All characters estimated simultaneously under Brownian Motion.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCV_dietaxa.png)

![\label{VCV_bestmodels} Best models (lowest AIC) supported in a pairwise character rate covariance analysis comparing correlated Brownian Motion models across the five selective regimes. Selective regimes were mapped onto the tree using an ancestral state reconstruction of the feeding guilds. Blank cells represent computationally singular contrasts.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCVbestModels.png)

![\label{VCV_ntaxa} Number of taxa used for each pairwise contrast in the VCV analyses, given the number of taxa without inapplicable states.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCV_ntaxa.png)

![\label{VCV_regimes} Pairwise estimated rate covariance matrices across the five selective regimes, using only taxa with diet data. Covariances scaled to correlations. Selective regimes were mapped onto the tree (22 species with diet data) using a stochastic mapping of the feeding guilds. Tree is pruned to taxa with no inapplicable states for a given character pair. Not all regimes are represented in all contrasts. Question marks represent computationally singular contrasts.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCV_regimes.png){height=600px}

![\label{VCV_diffs} Scaled differences between the regime-specific covariance matrices in \@ref{VCV_regimes} and the whole tree covariance matrix.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCVdiffs.png){height=600px}

![\label{VCV_diffs_anc} Scaled differences between the regime-specific covariance matrices in \@ref{VCV_regimes} and the covariance matrices in their preceding regime, the large-crustacean specialist regime.](Post-submission modifications/Supplementary_materials/Online_appendices/PNGs/VCVdiffs_anc.png)

 ~ ~ ~

## References {-}

